### DISCUSSION TITLE
How to resize partitions on startup with `systemd-repart`?

### DISCUSSION THREAD

**Discussion by paulmenzel** [systemd/mkosi#2654] [Q&A] [answered]
I want to create an virtual machine image for QEMU/KVM with a at least 100 GB root partition:

```
$ more mkosi.repart/*
::::::::::::::
mkosi.repart/00-esp.conf
::::::::::::::
[Partition]
Type=esp
Format=vfat
CopyFiles=/boot:/
CopyFiles=/efi:/
SizeMinBytes=512M
SizeMaxBytes=512M
::::::::::::::
mkosi.repart/05-bios.conf
::::::::::::::
[Partition]
# UUID of the grub BIOS boot partition which grubs needs on GPT to
# embed itself into.
Type=21686148-6449-6e6f-744e-656564454649
SizeMinBytes=1M
SizeMaxBytes=1M
::::::::::::::
mkosi.repart/10-root.conf
::::::::::::::
[Partition]
Type=root
Format=ext4
CopyFiles=/
Minimize=guess
```

(These are the default files.)

Adding *systemd-repart* to the package list, and building the image

    $ mkosi -d ubuntu -r jammy -t disk -o ubuntu-jammy-bigbluebutton --debug
    […]
    ‣  /home/build/src/ubuntu-bbb/mkosi.output/ubuntu-jammy-bigbluebutton.raw size is 1.5G, consumes 1.0G.
    […]

I convert it to qcow2, and resize it by adding 100 GB:

    $ qemu-img convert -f raw -O qcow2 ubuntu-jammy-bigbluebutton.raw ubuntu-jammy-bigbluebutton.qcow2
    $ qemu-img resize ubuntu-jammy-bigbluebutton.qcow2 +100G

Booting the image with QEMU and passing the Linux kernel directly

    $ qemu-system-x86_64 -M q35 -m 32G -enable-kvm -cpu host -smp cpus=32 -device virtio-rng-pci -net nic,model=virtio-net-pci -net user,hostfwd=tcp::22231-:22 -drive if=virtio,file=/scratch/local2/ubuntu-jammy-bigbluebutton.qcow2 -nographic -kernel /scratch/local2/vmlinuz-5.15.0-1057-kvm -append "root=/dev/vda3 debug init=/usr/bin/systemd"

the root partition was not expanded:

    root@localhost:~# lsblk
    NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
    sr0     11:0    1  1024M  0 rom
    vda    254:0    0 101.5G  0 disk
    ├─vda1 254:1    0   512M  0 part
    ├─vda2 254:2    0     1M  0 part
    └─vda3 254:3    0 971.3M  0 part /

For some reason, `systemd-repart.service` is not started:

    # systemctl --version
    systemd 249 (249.11-0ubuntu3.12)
    +PAM +AUDIT +SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
    # dpkg -l systemd-repart
    Desired=Unknown/Install/Remove/Purge/Hold
    | Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
    |/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
    ||/ Name           Version            Architecture Description
    +++-==============-==================-============-===================================
    ii  systemd-repart 249.11-0ubuntu3.12 amd64        Provides the systemd-repart utility
    # systemctl status systemd-repart
    systemd-journald[304]: Sent WATCHDOG=1 notification.
    ○ systemd-repart.service - Repartition Root Disk
         Loaded: loaded (/lib/systemd/system/systemd-repart.service; static)
         Active: inactive (dead)
      Condition: start condition failed at Tue 2024-04-23 22:03:10 UTC; 1min 50s ago
                 ├─ ConditionDirectoryNotEmpty=|/usr/lib/repart.d was not met
                 ├─ ConditionDirectoryNotEmpty=|/usr/local/lib/repart.d was not met
                 ├─ ConditionDirectoryNotEmpty=|/etc/repart.d was not met
                 ├─ ConditionDirectoryNotEmpty=|/run/repart.d was not met
                 ├─ ConditionDirectoryNotEmpty=|/sysroot/usr/lib/repart.d was not met
                 ├─ ConditionDirectoryNotEmpty=|/sysroot/usr/local/lib/repart.d was not met
                 ├─ ConditionDirectoryNotEmpty=|/sysroot/etc/repart.d was not met
                 ├─ ConditionDirectoryNotEmpty=|/sysusr/usr/lib/repart.d was not met
                 └─ ConditionDirectoryNotEmpty=|/sysusr/usr/local/lib/repart.d was not met
           Docs: man:systemd-repart.service(8)

The files from `mkosi.repart` are not in the image? Shouldn’t they be? Should the user do that manually again?

**Comment by daandemeyer** [comment, marked as answer]
@paulmenzel The files from mkosi.repart apply when mkosi builds the image. They do not become part of the image. You need to ship separate repart definition files in the image itself to have systemd-repart grow the partitions on first boot.

**Reply by paulmenzel** [reply]
Thank you for the confirmation.
